develop/2017-05-29-fgrehm-analysis-on-meals-based-on-presence-information.py

```diff
@@ -0,0 +1,467 @@
+            right_on=['congressperson_name', 'party', 'state'])
+print("Speeches with matching deputy:", len(speeches))
+
+
+# In[27]:
+
+speeches['first_speech_at'] = speeches['speech_started_at']
+speeches['last_speech_at'] = speeches['speech_started_at']
+speeches['total_speeches'] = 0
+aggregations = {
+    'last_speech_at': 'max',
+    'first_speech_at': 'min',
+    'total_speeches': 'count',
+}
+speeches = speeches.groupby(['congressperson_document', 'session_date'], as_index=False).agg(aggregations)
+print("Speeches grouped by date:", len(speeches))
+
+
+# In[29]:
+
+meals_outside_df_during_speeches = pd.merge(
+    left=speeches, 
+    right=meals_outside_df, 
+    left_on=['congressperson_document', 'session_date'],
+    right_on=['congressperson_document', 'issue_date']
+)
+print("Meals outside DF on days that the congressperson gave a speech:", len(meals_outside_df_during_speeches))
+
+meals_outside_df_during_speeches = meals_outside_df_during_speeches.query('company_name != "GOL LINHAS AEREAS S.A."')
+print("Meals excluding expenses on flight:", len(meals_outside_df_during_speeches))
+
+meals_outside_df_during_speeches = meals_outside_df_during_speeches[
+    ~meals_outside_df_during_speeches.company_name.str.contains('HOTEL')
+]
+print("Meals excluding expenses on hotels:", len(meals_outside_df_during_speeches))
+
+
+# In[34]:
+
+def score(meal):
+    return score_by_time(meal) + score_by_ts(meal)
+
+def score_by_time(meal):
+    if meal.total_speeches > 1:
+        return meal.first_speech_at.hour - meal.last_speech_at.hour + (meal.total_speeches - 1)
+    else:
+        return 0
+    
+def score_by_ts(meal):
+    if meal.timestamps == '':
+        return 0
+    
+    occurences = 0
+    for ts in meal.timestamps:
+        if ts == None:
+            continue
+        if (meal.first_speech_at.hour-1) < ts.tm_hour < (meal.last_speech_at.hour+1):
+            occurences += 1
+    return occurences * 10
+    
+suspects2 = meals_outside_df_during_speeches.copy()
+suspects2.timestamps.fillna('', inplace=True)
+suspects2['score'] = suspects2.apply(score, axis='columns')
+suspects2 = suspects2.query('score > 0').sort_values('score', ascending=False)
+print("Suspicious reimbursements:", len(suspects2))
+
+suspects2 = suspects2[~suspects2.document_id.isin(suspects.document_id)]
+print("Suspicious reimbursements that were not found using presences info:", len(suspects2))
+
+
+# In[35]:
+
+report(suspects2.head(20), [
+    'document_id',
+    'receipt',
+    'issue_date', 
+    'congressperson_name',
+    'score',
+    'company_state', 
+    'total_net_value', 
+    'first_speech_at', 
+    'last_speech_at',
+    'company_name'
+])
+
+
+# ## Meals in DF while the congressperson was outside DF
+# 
+# In the past we also found cases where the congressperson was on an official mission outside DF but requested a reimbursement for a meal in DF
+
+# In[36]:
+
+missions = pd.read_csv('../data/2017-04-19-official-missions.xz', low_memory=False)
```

Comment:
FTR - Looks like the chamber of deputies included a new column on their pages indicating whether a mission has been canceled ("Viagem Cancelada" in http://www.camara.leg.br/missao-oficial/missao-pesquisa?deputado=1&nome-deputado=&nome-servidor=&dati=06%2F03%2F2017&datf=04%2F06%2F2017&nome-evento=), I've regenerated the dataset and seems that it got rid of quite a few suspicious reimbursements